typescript/infra/scripts/verify-validator.ts

```diff
@@ -0,0 +1,297 @@
+    c: controlBucket,
+  } = await getArgs();
+
+  const cClient = new S3Wrapper(controlBucket);
+  const pClient = new S3Wrapper(prospectiveBucket);
+
+  const [{ obj: cLatestCheckpoint }, { obj: pLastCheckpoint }] =
+    await Promise.all([
+      cClient.getS3Obj<number>('checkpoint_latest_index.json').catch((err) => {
+        console.error(
+          "Failed to get control validator's latest checkpoint.",
+          err,
+        );
+        process.exit(1);
+      }),
+      pClient.getS3Obj<number>('checkpoint_latest_index.json').catch((err) => {
+        console.error(
+          "Failed to get prospective validator's latest checkpoint.",
+          err,
+        );
+        process.exit(1);
+      }),
+    ]);
+
+  console.assert(
+    Number.isSafeInteger(cLatestCheckpoint),
+    'Expected latest control checkpoint to be an integer',
+  );
+  console.assert(
+    Number.isSafeInteger(pLastCheckpoint),
+    'Expected latest prospective checkpoint to be an integer',
+  );
+
+  console.log(`Latest Index`);
+  console.log(`control: ${cLatestCheckpoint}`);
+  console.log(`prospective: ${pLastCheckpoint}\n`);
+
+  let extraCheckpoints = [];
+  const missingCheckpoints = [];
+  let invalidCheckpoints = [];
+  const modTimeDeltasS = [];
+  const fullyCorrectCheckpoints = [];
+  let missingInARow = 0;
+  let lastNonMissingCheckpointIndex = Infinity;
+  for (let i = Math.max(cLatestCheckpoint, pLastCheckpoint); i >= 0; --i) {
+    if (missingInARow == MAX_MISSING_CHECKPOINTS) {
+      missingCheckpoints.length -= MAX_MISSING_CHECKPOINTS;
+      invalidCheckpoints = invalidCheckpoints.filter(
+        (j) => j < lastNonMissingCheckpointIndex,
+      );
+      extraCheckpoints = extraCheckpoints.filter(
+        (j) => j < lastNonMissingCheckpointIndex,
+      );
+      break;
+    }
+
+    const key = `checkpoint_${i}.json`;
+
+    let c: Checkpoint | null;
+    let cLastMod: Date | null;
+    try {
+      const t = await cClient.getS3Obj(key);
+      if (isCheckpoint(t.obj)) {
+        if (t.obj.checkpoint.index != i) {
+          console.log(`${i}: Control index is invalid`, t);
+          process.exit(1);
+        }
+        [c, cLastMod] = [t.obj, t.modified];
+      } else {
+        console.log(`${i}: Invalid control checkpoint`, t);
+        process.exit(1);
+      }
+    } catch (err) {
+      c = cLastMod = null;
+    }
+
+    let p: Checkpoint;
+    let pLastMod: Date;
+    try {
+      const t = await pClient.getS3Obj(key);
+      if (isCheckpoint(t.obj)) {
+        [p, pLastMod] = [t.obj, t.modified];
+        lastNonMissingCheckpointIndex = i;
+      } else {
+        console.log(`${i}: Invalid prospective checkpoint`, t.obj);
+        invalidCheckpoints.push(i);
+        continue;
+      }
+      if (!c) {
+        extraCheckpoints.push(i);
+      }
+      missingInARow = 0;
+    } catch (err) {
+      if (c) {
+        missingCheckpoints.push(i);
+        missingInARow++;
+      }
+      continue;
+    }
+
+    console.assert(
```

Comment:
may be easier to read this loop if we pulled out these asserts into some function intented to assert the "correctness" of a checkpoint.

Generally just observing from taking a second look and getting my bearings again that future readers / editors could benefit from pulling some stuff out of here

Reply:
Oh, I think I see now